<a href="https://colab.research.google.com/github/jjcrofts77/TMB-MATH34041/blob/main/content/notebooks/Chapter5/PythonBasics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 0.1 Basics in Python

This section aims to cover simple, core programming concepts in Python.

## Installing Python and add-ons
If you wish to install Python on your own computer it is advisable to use an integrated system (IS) to avoid compatability issues. For this course the recommended IS is [Anaconda](https://www.anaconda.com/), which is free to use and ships with [Spyder](https://www.spyder-ide.org/) a scientific Python development environment (see Figure {numref}`FigSpyderIDE3` for an example of the Spyder environment).

```{figure} ../../images/SpyderIDEExample.png
---
height: 450px
name: FigSpyderIDE3
---
Illustration of a Python script as seen in the Anaconda Spyder IDE
```

Once installed, you can launch Spyder through the Anaconda Navigator, which can be found under the list of programs prompted by opening Windows Start menu.

The IDE consists of an Editor Window, an Object Inspector Window and The Console. The Editor Window is where we write our Python scripts; the Object Inspector is where we browse files, explore variables in the workspace and visulise plots; whilst the Console is where we access the iPython shell and see the results of our scripts.

To create a Python script we enter commands into the editor, save our progam and run it by clicking on the green triangle in the upper-lefthand corner of the IDE; the results will displayed in the Python console in the lower-righthand pane of the IDE. (Again, see Figure {numref}`FigSpyderIDE3`)

## Google Colab

Google Colab, or Colaboratory, is a free Jupyter notebook environment that runs in the cloud. It allows you to write and execute Python code, as well as other programming languages, in a web browser. Colab notebooks are a great way to combine code, text, and images in a single document. This makes it easy to document your work and share it with others.

Colab has two types of cells: text and code. Text cells are formatted using a simple markup language called Markdown.

See the following link for a brief description of the [Google Colab Markdown](https://colab.research.google.com/notebooks/markdown_guide.ipynb#scrollTo=Lhfnlq1Surtk) langauge.

## Python as a calculator
Python can be used as a calculator to perform basic and advanced arithmetic operations. Here are some examples of how to use Python as a calculator:

  **Basic arithmetic operations:** Python can be used to perform basic arithmetic operations such as addition, subtraction, multiplication, and division. For example, the following code would add two numbers and print the result:

In [ ]:
a = 10
b = 5

print(a + b)

15


  **Advanced arithmetic operations:** Python can also be used to perform more advanced arithmetic operations such as exponentiation, logarithms, and trigonometric functions. For example, the following code would calculate the square root of a number and print the result:

In [ ]:
import math

number = 16

print(math.sqrt(number))

4.0


In the above we have imported the Python math module which contains the standard mathematical functions (*e.g.* $\cos, \sin, \exp, \sinh, \cosh$ *etc.*).


# Basic programming in Python

In this section we shall see how Python can be used to write simple Python programs. We shall focus on three programming structures:

  1. defining functions
  2. using for and while loops
  3. if, elif, else constructs

We have seen versions of these structures when learning Matlab in Years 1 and 2.

## Defining functions
Below I have written a function for converting degrees Celsius to Kelvin.


In [ ]:
# A function to convert degrees Celsius to Kelvin

def C2K():
  C = int(input('Enter temperature in degrees Celsius: '))
  K = C + 273.15
  print('Temperature in Kelvin is {} K'.format(K))

# run the function
C2K()

Note that the function above did not require any inputs.

Suppose we want to write a program that defines the *logistic equation* given by

$$f(x) = \mu x(1-x)$$

Then we would define a function as follows

In [3]:
# The logistic equation - save as f_mu.py

"""
Created on Wednesday July 12th 14:29
@author: JJCrofts
"""
def f_mu(mu, x):
  return mu*x*(1-x)

# run the function for a given input
mu, x = 1, 0.25
fm = f_mu(mu, x)
print('fm = {}'.format(fm))

fm = 0.1875


## Using loops
The following examples illustrate how for and while loops can be used to speed up repetitive tasks.

Let us write a program that list the first $n$ terms of the Fibonacci sequence

$$
x_{k+2} = x_{k+1}+x_k: \quad x_0 = 0, x_1 = 1
$$


In [ ]:
# A function to list the first n terms of the Fibonacci sequence

def fibonacci(n):
  a, b = 0, 1
  print(a)
  print(b)
  print(a+b)
  for i in range(n-3):
    a, b = b, a+b
    print(a+b)

# run function
fibonacci(20)

Recall that the above Fibonacci sequence can be solved analytically to obtain

$$
x_k = \frac{\phi^k-\psi^k}{\sqrt{5}}
$$

where

$$
\phi = \frac{1+\sqrt{5}}{2}
$$

is the *golden ratio* and $\psi=1-\phi$ is its conjugate.

We can check our result using this formula. For example, for $n=20$ we have that

$$
x_{20} = \frac{\phi^{20}-\psi^{20}}{\sqrt{5}} = 4181
$$

which agrees with our Python code.

Next lets use loops to sum the first $n$ natual numbers

In [2]:
# A function to sum the first n natural numbers

def sum_n(n):
  sum = 0
  i = 1
  while i <= n:
    sum += i
    i += 1   # increment counter
  print('The sum is', sum)

# run function
sum_n(100)

The sum is 5050


Again this is easy to check since we know that

$$
\sum_{i=1}^{100} = \frac{100(101)}{2} = 5050
$$

as expected.

## Conditional statements using if, elif, else

Lets write a program that returns a students grade for the Maths Bio coursework given a score out of 100.

In [4]:
# A program to grade the Topics in Mathematical Biology coursework

def grade(score):
  if score >= 92.5:
    GBA = '1EXC'
  elif score >= 84.5:
    GBA = '1HIGH'
  elif score >= 77.5:
    GBA = '1MID'
  elif score >= 69.5:
    GBA = '1LOW'
  else:
    GBA = 'Fail'
  return GBA

grade(94)

'1EXC'

What about a game of guess the number!

In [ ]:
# Guess the number

import random # Import the random number module

num_guesses = 0
name = input('Hi! What is your name? ')
number = random.randint(1, 20) # a random number between 1 and 20
print('Welcome, {}! I am thinking of an integer between 1 and 20.'.format(name))

while num_guesses < 6:
  guess = int(input('Take a guess and type the integer '))
  num_guesses += 1

  if guess < number:
    print('Your guess is too low')
  if guess > number:
    print('Your guess is too high')
  if guess == number:
    break

if guess == number:
  print('Damn {}, you win! You guessed the correct number in {} guesses'.format(name, num_guesses))
else:
  print('You lose! The number I was thinking of was {}'.format(number))

In the above program we have imported a module in order to access functions for simulating random numbers. In order to use Python to study networks we are going to have to import other modules and libraries as will be demonstrated in the sections to follow.